### PR TITLE
Enhance squashfs extraction handling.

### PR DESCRIPTION
**Pull request by QKaiser**
We've moved to our own version of squashfs-tools to be able to use the '-no-exit-code' command line option. This option makes unsquashfs not set the exit code to zero on non-fatal errors. non-fatal errors include the inability to create special files  such as block devices. With this option, unblob will not report an error when such files cannot be created due to missing super user privileges.

This is an opinionated change that follows the behavior of JFFS2 (jefferson) and RomFS (unromfs) extractors that simply do not attempt to create special files if the process lacks super user privileges.

Dockerfile and github actions have been updated to include the installation of our own fork of squashfs-tools.

An squashfs integration test files containing block device definition has been added to check that unblob does not report an error.

### CODE REVIEW

**Comment by QKaiser** [comment]
@martonilles could you move your squashfs-tools (https://github.com/martonilles/squashfs-tools) fork under IoT Inspector organization ? I'll then edit this PR with the new link.

**Comment by QKaiser** [comment]
We should not merge this until support for other compression formats is added to https://github.com/martonilles/squashfs-tools like it's done in sasquatch. At this point it only support gzip and missing lzma, lzma-adaptive, lzma-alt, lzma-ddwrt, lzo, and xz.

**Comment by martonilles** [comment]
I moved sasquatch to https://github.com/IoT-Inspector/sasquatch
Also rebased it on squashfs-tools 4.5 and latest master, you can find it in the 4.5-sasquatch and sasquatch branches

**Comment by QKaiser** [comment]
> I moved sasquatch to https://github.com/IoT-Inspector/sasquatch Also rebased it on squashfs-tools 4.5 and latest master, you can find it in the 4.5-sasquatch and sasquatch branches

Can you make it public so we can use it in Dockerfile and github actions ?

**Comment by martonilles** [comment]
> > I moved sasquatch to https://github.com/IoT-Inspector/sasquatch Also rebased it on squashfs-tools 4.5 and latest master, you can find it in the 4.5-sasquatch and sasquatch branches
> 
> Can you make it public so we can use it in Dockerfile and github actions ?
@QKaiser  yes, done

**Comment by vlaci** [comment]
I have created a [branch](https://github.com/IoT-Inspector/unblob/commits/nix-sasquatch) for the nix stuff to use the new version. Keep in mind, that the latest squashfs-tools in nix already has the `-no-exit-code`. Also, tests using this fork fail for me:
```
  python3.9-unblob> 2022-02-23 13:52.47 [error    ] Extract command failed         command=unsquashfs -f -d /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.nopad.bin_extract/0-320.squashfs_v3_extract /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.nopad.bin_extract/0-320.squashfs_v3 exit_code=0x-6 handler=squashfs_v3 pid=1657 stderr=SquashFS version [768.256] / inode count [83886080] suggests a SquashFS image of a different endianess
python3.9-unblob> Reading a different endian SQUASHFS filesystem on /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.nopad.bin_extract/0-320.squashfs_v3
python3.9-unblob> *** buffer overflow detected ***: terminated
python3.9-unblob>  stdout=
python3.9-unblob> 2022-02-23 13:52.47 [error    ] Extract command failed         command=unsquashfs -f -d /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.4kpad.bin_extract/0-4096.squashfs_v3_extract /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.4kpad.bin_extract/0-4096.squashfs_v3 exit_code=0x-6 handler=squashfs_v3 pid=1658 stderr=SquashFS version [768.256] / inode count [83886080] suggests a SquashFS image of a different endianess
python3.9-unblob> Reading a different endian SQUASHFS filesystem on /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.4kpad.bin_extract/0-4096.squashfs_v3
python3.9-unblob> *** buffer overflow detected ***: terminated
python3.9-unblob>  stdout=
python3.9-unblob> 2022-02-23 13:52.47 [error    ] Extract command failed         command=unsquashfs -f -d /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.1kpad.bin_extract/0-1024.squashfs_v3_extract /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.1kpad.bin_extract/0-1024.squashfs_v3 exit_code=0x-6 handler=squashfs_v3 pid=1659 stderr=SquashFS version [768.256] / inode count [83886080] suggests a SquashFS image of a different endianess
python3.9-unblob> Reading a different endian SQUASHFS filesystem on /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3.be.1kpad.bin_extract/0-1024.squashfs_v3
python3.9-unblob> *** buffer overflow detected ***: terminated
python3.9-unblob>  stdout=
python3.9-unblob> 2022-02-23 13:52.47 [error    ] Extract command failed         command=unsquashfs -f -d /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3_be.bin_extract/0-4096.squashfs_v3_extract /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3_be.bin_extract/0-4096.squashfs_v3 exit_code=0x-6 handler=squashfs_v3 pid=1656 stderr=SquashFS version [768.0] / inode count [50331648] suggests a SquashFS image of a different endianess
python3.9-unblob> Reading a different endian SQUASHFS filesystem on /build/pytest-of-nixbld/pytest-0/test_all_handlers_filesystem_s1/squashfs_v3_be.bin_extract/0-4096.squashfs_v3
python3.9-unblob> *** buffer overflow detected ***: terminated
python3.9-unblob>  stdout=
```

**Comment by QKaiser** [comment]
I'm observing the same errors with squashfs v3 big endian sample.

**Comment by martonilles** [comment]
Fixed squashfs segfault here: https://github.com/IoT-Inspector/sasquatch/commit/b5939b3ef0db8dbd9e4203ff9112baf77cdd0449

**Comment by vlaci** [comment]
Confirmed, it works better.

**Comment by vlaci** [comment]
Because of the missing sasquatch binary

**Comment by vlaci** [comment]
I am a bit concerned about building sasquatch from scratch for each and every revision of every pull request...

**Comment by QKaiser** [comment]
> I am a bit concerned about building sasquatch from scratch for each and every revision of every pull request...

I guess we could build and release statically linked binaries from sasquatch repo.

**Comment by QKaiser** [comment]
@vlaci please check that everything is right for nix packaging

File: .github/actions/setup-dependencies/action.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ runs:
   using: "composite"
   steps:
     - name: Install 3rd party from apt
-      run: sudo apt install unar zlib1g-dev liblzo2-dev lzop lziprecover img2simg
+      run: sudo apt install gcc img2simg liblzo2-dev lziprecover lzop unar zlib1g-dev
```
[the remark this answers is not in the document]
Comment (QKaiser) [reply .github/actions/setup-dependencies/action.yml:10]: Yes we need it to install yara-python :)

[the remark this answers is not in the document]
Comment (QKaiser) [reply .github/actions/setup-dependencies/action.yml:10]: That's what happens in docker build without gcc:

```
Building wheel for yara-python (setup.py): finished with status 'error'
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [21 lines of output]
      /usr/local/lib/python3.8/site-packages/setuptools/dist.py:697: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
        warnings.warn(
      running bdist_wheel
      running build
      running build_ext
      building 'yara' extension
      creating build
      creating build/temp.linux-x86_64-3.8
      creating build/temp.linux-x86_64-3.8/yara
      creating build/temp.linux-x86_64-3.8/yara/libyara
      creating build/temp.linux-x86_64-3.8/yara/libyara/proc
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules/time
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules/pe
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules/math
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules/elf
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules/demo
      creating build/temp.linux-x86_64-3.8/yara/libyara/modules/tests
      gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -D_GNU_SOURCE=1 -DUSE_LINUX_PROC=1 -Iyara/libyara/include -Iyara/libyara/ -I. -I/usr/local/include/python3.8 -c yara-python.c -o build/temp.linux-x86_64-3.8/yara-python.o -std=c99
      unable to execute 'gcc': No such file or directory
      error: command 'gcc' failed with exit status 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for yara-python

```

[the remark this answers is not in the document]
Comment (vlaci) [reply .github/actions/setup-dependencies/action.yml:10]: But we would have need that before as well, I guess? There is no change here in how we install yara-python at all.

[the remark this answers is not in the document]
Comment (QKaiser) [reply .github/actions/setup-dependencies/action.yml]: Looks okay !

**Review by vlaci — APPROVED** [review]

**Pull request merged**